webauthn-rs-core/src/attestation.rs

```diff
@@ -17,6 +17,86 @@ use openssl::stack;
 use openssl::x509;
 use openssl::x509::store;
 use openssl::x509::verify;
+use x509_parser::oid_registry::Oid;
+
+/// x509 certificate extensions are validated in the webauthn spec by checking
+/// that the value of the extension is equal to some other value
+pub(crate) trait AttestationX509Extension {
```

Comment:
Can you implement this for anonca (look for `1.2.840.113635.100.8.2`) because I'm wondering if this trait will "feel awkward" to use, and that's a good test.

Reply:
I had mocked this up at one point. I think the following works (if you add the new `WebauthnError` variants that I made up below)
```rust
impl AttestationX509Extension for AppleAnonymousNonce {
    type Output = [u8; 32];

    // 4. Verify that nonce equals the value of the extension with OID ( 1.2.840.113635.100.8.2 ) in credCert. The nonce here is used to prove that the attestation is live and to protect the integrity of the authenticatorData and the client data.
    const OID: Oid<'static> = der_parser::oid!(1.2.840 .113635 .100 .8 .2);

    fn parse(i: &[u8]) -> der_parser::error::BerResult<&Self::Output> {
        use der_parser::{der::*, error::BerError};
        parse_der_container(|i: &[u8], hdr: Header| {
            if hdr.tag() != Tag::Sequence {
                return Err(nom::Err::Error(BerError::BerTypeError.into()));
            }
            let (i, tagged_nonce) = parse_der_tagged_explicit(1, parse_der_octetstring)(i)?;
            let (class, _tag, nonce) = tagged_nonce.as_tagged()?;
            if class != Class::ContextSpecific {
                return Err(nom::Err::Error(BerError::BerTypeError.into()));
            }
            let nonce = nonce
                .as_slice()?
                .try_into()
                .map_err(|_| der_parser::error::BerError::InvalidLength)?;
            Ok((i, nonce))
        })(i)
    }

    fn if_missing() -> Result<(), WebauthnError> {
        Err(WebauthnError::AttestationStatementMissingExtension)
    }

    fn if_unequal() -> WebauthnError {
        WebauthnError::AttestationCertificateNonceMismatch
    }
}
```

Reply:
Want to add it to the PR? I think that example has shown me I was wrong to doubt you :)